### ISSUE TITLE
Password protected livestream

### ISSUE THREAD

**Issue by aaronpk** [owncast/owncast#489]
Hello! I'd love to be able to create a password-protected livestream. I'm imagining the simplest possible ui for this, which is just to prompt the visitor to enter the password when they first land on the web page. The primary use case would be to provide members-only livestreams for Patreon/etc, or private group streams for things like family video watch parties. In particular that means:

* User accounts are not needed, no individual passwords
* No direct link to prefill the password (prevents sharing the direct link, but if users want to share the password around that's an acceptable risk)

The goal is not to make a perfectly secure livestream viewer, just to prevent casual link sharing and public indexing of the content.

I am planning on setting this up in the mean time by putting owncast behind an [nginx proxy that requires authentication](https://github.com/vouch/vouch-proxy), but that's definitely more steps than most people should have to do to accomplish this.

**Comment by vilmibm** [comment]
FWIW, I set up via the Droplet thing on Digital Ocean and was able to add a `basicauth / { ... }` directive to `/etc/Caddyfile`; obviously not much help though if you are not running via Caddy/the droplet helper but if you are it was an easy thing to add.

**Comment by DaleJP** [comment]
I know this is a place for pros so I"m sorry to muddy it, but I reaaaaally need to password protect my stream for my small international school "TV Channel" as the children will be featured on it and the above comment is Greek to me and I do not understand it.
I just need to simply have a password unlock the stream for the end user that I can distribute to the families.
I can embed the player on a password protected page on our website but the URL will still be exposed to anyone with previous access and anything that happens upon the URL.

Am I just out of luck? Or do you have any suggestions for me? Feel free to remove this comment if inappropriate.

**Comment by gabek** [comment]
@DaleJP Are you using a HTTP proxy in front of Owncast to support SSL? If so, we can tell you how to configure it to require authentication.

**Comment by DaleJP** [comment]
@gabek Thank you. I haven't set it up yet until I can figure out how to password protect it.
I have NextCloud currently running on the server I hope to install Owncast on with Apache and SSL from Let'sencrypt if that is of any use.

**Comment by gabek** [comment]
Absolutely. If you get Apache running as a SSL proxy in front of Owncast it should be pretty straightforward to tell apache to also request authentication.  While I've never done it with Apache myself, and it's probably the most complex option, there's no reason it shouldn't work.

https://owncast.online/docs/sslproxies/apache/ + https://stackoverflow.com/questions/724599/setting-up-an-apache-proxy-with-authentication might put you in the right direction. But if Apache isn't required I _highly_ recommend Caddy instead, you'll get where you're going much easier.

<blockquote><img src="/images/owncast-logo-1000x1000.png" width="48" align="right"><div><img src="/favicon.png" height="14"> Owncast</div><div><strong><a href="/docs/sslproxies/apache/">Apache</a></strong></div><div>If you're already using Apache you can use it as a proxy.</div></blockquote>
<blockquote><img src="https://cdn.sstatic.net/Sites/stackoverflow/Img/apple-touch-icon@2.png?v=73d79a89bded" width="48" align="right"><div>Stack Overflow</div><div><strong><a href="https://stackoverflow.com/questions/724599/setting-up-an-apache-proxy-with-authentication">Setting up an Apache Proxy with Authentication</a></strong></div><div>I need to set up a proxy with authentication to verify the behavior of an application that connects to the internet.

I am trying to set-up an Apache installation with forward proxy and authenticat...</div></blockquote>

**Comment by DaleJP** [comment]
@gabek Brilliant. Thank you. 
I'll look into Caddy if it is the preferred method but I would also like to understand Apache a bit better. It's a magical box to me at the moment as I've only ever just copied and pasted from tutorials.

Thanks a lot. I think this should get me on track.

**Comment by vmirinav** [comment]
@gabek If I setup Apache proxy with authentication will the <iframe embedding work still it cant do POST to login to the server ?

**Comment by gabek** [comment]
You could setup your proxy to operate any way you like, but the default answer will be no. If you lock down your Owncast server, then you locked down your Owncast server. But you could configure your proxy to exclude certain paths if you wanted, though I'm not sure why you'd want to make your stream password protected and completely open to embeds at the same time. But I'm not here to judge :)

Keep in mind I've never used any of these auth proxies myself, so you'll have to dig into their documentation to configure them.

**Comment by JeremyMahieu** [comment]
Would also be interesting if the password can be included in the link. For example `http://my.owncast.net/?p=dfiAd21krgQ1fdga` which then creates a jwt token en redirects to http://my.owncast.net/. This way there's no real UI you need to go through.

The idea is only people with the link can view the stream and it's not just discoverably by having the domain name.

**Comment by dokgu** [comment]
I have the same use case as @aaronpk - I would really love to have a private livestream.

**Comment by gabek** [comment]
@dokgu See https://github.com/owncast/owncast/discussions/649

<blockquote><img src="https://opengraph.githubassets.com/ba4c3787f0115f748e4485d712303ed11e3ff1c0c025a955dab02636c20c30ce/owncast/owncast/discussions/649" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/owncast/owncast/discussions/649">Best way to host "secret" streams? · owncast/owncast · Discussion #649</a></strong></div><div>We're evaluating Owncast to host, among other events, online movie nights for a group of friends. For this purpose it would be nice to have Viewer authentication, an access password for watching a ...</div></blockquote>

**Comment by gabek** [comment]
So here's the deal about Owncast supporting "Password protected streams".

One person might want a single basic password.
Another might want a different password for each person, requiring user management.
Others would want them to use their Discord password.
Another would want them to login with Twitter.
Others to use the password they use on a custom website.
Etc etc. It goes on forever.

So that's why each person setting up their own auth, either with something like an OAuth proxy (Vouch), or their own custom auth layer makes the most sense. It's simply impossible to do this within Owncast itself.

If there was a good way for Owncast itself to support this, then we could do it, but there's an infinite number of ways "password protected" could mean for different people.

**Comment by Sythelux** [comment]
I want to emphasize, that password protected livestreams would be an important feature and although it would be cool to have a whole system with multiple and different authentication solutions, the development of such would probably take some time.

A simple implementation of a one time password would help everyone for now and could be done rather soon and can later be deprecated, when there is a proper API without hurting anyone.

**Comment by gabek** [comment]
>  it would be cool to have a whole system with multiple and different authentication solutions

That's precisely what I was trying to say above, such a thing would not be possible. What "multiple"? The options are limitless. 

The options you think would be reasonable to build directly into Owncast might be terrible ideas for somebody else. Maybe the US Government uses Owncast, should I build in support for the US government authentication services if they ask? Everyone likes TikTok today, it would make sense to package Owncast with the TikTok SDK then to support their authentication, right?

> A simple implementation of a one time password would help everyone for now and could be done rather soon and can later be deprecated, when there is a proper API without hurting anyone.

What proper API? I feel like there's a lot of blanks that are being skipped over. And it sounds like you're asking me to build some kind of authentication plugin system, and then asking every admin that runs Owncast who wants to take advantage of it to write custom plugins? Because you'd rather not use a reverse proxy?

There is no good way to do this, that is:

1. Flexible, so any kind of auth that anybody can think of can be used.
2. Approachable so the admin would be able to add that auth themselves without requiring me at all, and without requiring them to be a software engineer.
3. Unopinionated so any style of auth is possible, and isn't leaving anybody out, from the enterprise users, to the social and casual streamers.

If you had a specific solution in mind, I'd love to hear it. Maybe something regarding auth standards?

But the best solution is the solution people have already been using. This is one of the roles a reverse web proxy in front of your server is good for, so you can gate your server in any way you can think of. If you wanted to run a little web server to serve up some files, but wanted them to login with Facebook first, how would you go about that? Would you tell the web server project to add Facebook support to the server directly? Of course not. You'd build your own thing that suits your needs using the tools that already exist in the world. This is no different in any way.

**Comment by aaronpk** [comment]
I keep getting pinged on this when anyone comments on this, so I figured I'd chime in again 3 years later 😂

I've had success putting Owncast behind a reverse proxy, then adding authentication options in the reverse proxy using various methods. This is the best option for handling things like enterprise SSO, login with GitHub, login with Discord/Twitch/Patreon etc. This is better than Owncast building in OAuth/OIDC support for the various providers directly because:

* It can support arbitrary providers that Owncast doesn't know about and that there isn't a good Go library for
* You can reuse the reverse proxy configuration for protecting other websites with the same SSO

With this in mind, I do think there is still a use for the simple password-protected livestream I outlined in the original issue.

* A single password for the livestream
* No user accounts, no individual passwords
* No direct link to prefill the password (prevents sharing the direct link, but if users want to share the password around that's an acceptable risk)

I feel like you can draw a firm line with password protected streams and not add any functionality beyond a single password as described here, and offload all other user authentication issues to a reverse proxy solution. I think there are plenty of interesting uses of a simple shared password for a livestream where configuring a reverse proxy is far too much work for the admin.

**Comment by Sythelux** [comment]
@gabek Apologies. Ignore my post above actually. I somehow misinterpreted a few things in here. So the official suggestion is "please use a proxy, that can do authentication for the level you want?" Then that is totally fine. I have this issue here subscribed for quite a while in the hope it will eventually be closed and a Soluton presented.
So is that the proxy thing a valid and working solution? in this case I'd say:
Write in the documentation, that this the current solution. Maybe give some hints there on which proxies work, don't work etc. and close this ticket here?

**Comment by gabek** [comment]
My concern about the single password, is it's perhaps too limiting in reality. It sounds like it's good enough, for simple use cases.

But I do feel even on one day of it in use you'd run into the password being leaked, and you'd have the option then of changing it and locking everyone else out, or just letting it leak. And the next logical question would be "the password leaked, could we just have a user system so we could manage users and passwords?" A reasonable question given the limitations and the problems of that particular solution.

One person did post somewhere how they added their own http basic auth for multiple users in front of their stream and that seemed to be both a relatively simple and low lift solution, especially if you already are using a proxy. I wonder if there's a middle ground there.

**Comment by aaronpk** [comment]
Yeah it's tough. I would say if you are trying to do something like members-only streams or sharing a livestream with a large group of people, you're better off with a proper user authentication system and a reverse proxy. The single password is for things like:

* I want to do some test streams and want to make sure nobody else can see them
* I'm going to stream the cameras in my home while I'm on vacation and let my 2 cat sitters watch the stream
* Family movie watch parties

Basically the recurring theme here is that the number of people who would have the password is extremely limited. You're totally right that trying to use it for a wider audience opens up a whole bunch of new problems. Maybe the answer is that it's still a bad idea because people will use the password feature for a large audience even if the docs say they shouldn't.

**Comment by Sythelux** [comment]
Ok so from what I found is. Nginx has support for OAth, Vouching, BasicHTTP-Auth through plugins or even built in. So it would be totally valid for the software owner of this repo to say. "for proper auth use your favourite proxies auth" and leave that to the server admin to do, that would be fine. You'd just need to keep owncast to be able to run behind a proxy in the future and we are good to go.

So for me I'd probably read into how to do vouching. Another Admin can try to setup nginx with oauth if needed and that would be totally fine I think. 

My goal was really to have a server running for a very small group of ppl. who can through around some one time passwords that just stop to working on their own after like 10minutes

(also sorry for using this issue board for discussing this now even though there is a discussion for it)

**Comment by gabek** [comment]
I think documentation around all the different options would be a nice addition. Not necessarily in detail, since there's a lot, but simply listing some choices and what they do might be better than the nothing that's provided right now.

**Comment by DaleJP** [comment]
Forgive my incompetence, but I'm a nursery principal, and reverse proxies are the biggest pain in the neck I have had when setting up interesting things for my school. I could never get them to work correctly, and I'm not confident they are set up properly, even now😅.
I wanted to be able to fire up one piece of software for events and observation days where the parents could just access it in a browser with a simple password and sometimes we want to do it through a mobile internet connection. 
It would be great if Owncast provided a simple password for streams to keep it simple.
Unfortunately, many of our parents here in Japan are not tech-savvy at all and anything beyond clicking a link and typing a password causes them to give up. We currently use YouTube and half of the parents just don't view it as they don't have Google accounts and don't want one.

**Comment by gabek** [comment]
@DaleJP I totally understand. But see above why I don't believe a single simple password would be a solution I would want to offer. A bad solution, in my opinion, is not better than no solution at all. All it would take is one person sharing that password either accidentally, or on purpose, and you would have the children's events that you thought were "private" not being private at all. And in the case you discovered the password got out, and you changed it, all of your existing viewers would get locked out until you got the new password to them. This would repeat every single time this happened.

I would not feel comfortable offering this option because I know how this would play out.

At its core, Owncast is a live video streaming server. I know it would be great if it were also an access control server, and user management, and everything that we can think of all in one. But that just isn't viable. It already does so much.

If there were a reasonable option that addressed most people who have this need, all at once, I'd certainly be up for talking about what that is. But right now, it's either too simple of a solution, or too complex of a solution. I continue to be all ears and open to any specific idea somebody might have, however.

**Comment by aaronpk** [comment]
Yeah this is a good example of when a proper access control mechanism is needed, and when a simple password for this use case is likely to cause massive problems at some point in the future.

**Comment by DaleJP** [comment]
I understand. It's your house, and you've made a great piece of software. 

I've recently figured out Cloudflare Zero Trust for stuff at our school, and it seems to be working well, but I need to figure out something that the parents with no accounts/email can do and something mobile. They only have emails their phone providers give them.

We never stream anything sensitive, but at least one layer of protection is nice, just in case.

**Comment by mm86133** [comment]
Just keep it KISS, you seem to overthink very simple issue that is solved by everyone else.

If it is streaming, SaaS or anything else.

Start with looking into access matrix and simple implementation would be sufficient.

Application should not be relying on extarnal factors. Follow SOLID. That is the best advice for the future.

**Comment by MagicLike** [comment]
Any updates on this? This would make owncast perfect for a usecase of mine

**Comment by gabek** [comment]
@MagicLike The above thread didn't result in any planned roadmap items. Could you be more specific about what updates you're looking for?

**Comment by MagicLike** [comment]
> @MagicLike The above thread didn't result in any planned roadmap items. Could you be more specific about what updates you're looking for?

I am looking for a way to make a stream which will only be vieawable for specific people who authorise thenselves via a password or what would be my preferred solution: a username and a password (so basically basic http auth), so that multiple users have their own credentials to access the stream. I could make it work using the integrated features of my reverse proxy and make it actually work using basic http auth, but it is not the most elegant, beautiful nor most secure solution. (afaik NGINX only uses MD5 for hashing the passwords)
Therefore an integrated solution like this would be really great.

**Comment by gabek** [comment]
That is not something that's in the roadmap, as everyone has their own requirements and opinions. There is no integrated solution that would fit for everyone. That's why the recommendation is for you to use a reverse proxy, so you can have what you need. Nobody is saying you have to use Nginx if you don't believe it's secure enough. https://github.com/owncast/owncast/discussions/649 has other suggestions, as well.

<blockquote><img src="https://opengraph.githubassets.com/ba4c3787f0115f748e4485d712303ed11e3ff1c0c025a955dab02636c20c30ce/owncast/owncast/discussions/649" width="48" align="right"><div><img src="https://github.githubassets.com/favicons/favicon.svg" height="14"> GitHub</div><div><strong><a href="https://github.com/owncast/owncast/discussions/649">Best way to host "secret" streams? · owncast owncast · Discussion #649</a></strong></div><div>We're evaluating Owncast to host, among other events, online movie nights for a group of friends. For this purpose it would be nice to have Viewer authentication, an access password for watching a ...</div></blockquote>

**Comment by shrik3** [comment]
Just some notes on nginx basic http auth:

since the admin auth also uses HTTP basic auth, there will be conflicts if you simply set a password globally like below.

```nginx
location / {
    proxy_pass http://localhost:8888;
    # other proxy related stuffs.   
    auth_basic "some text ...";
    auth_basic_user_file /etc/nginx/htpasswd;
}
```

And obviously, while I'm sharing a live with some friends, I don't want to give them admin privileges, i.e. the viewer login and admin login must be different.

I tried to turn off nginx's http auth for `/admin/` but the admin pages also use `/api/` and various other routes, which are protected with the viewer password.

My current "workaround" is to only protect the stream (i.e. `/hls/`):

```nginx
location / {
    proxy_pass http://localhost:8888;
    # settings.
}

location /hls/ {
    proxy_pass http://localhost:8888;
    # settings.
    auth_basic "some text ...";
    auth_basic_user_file /etc/nginx/htpasswd;
}
```

I'm not so experienced with nginx magics ... I'll appreciate it if you could showcase how to set this up properly.

**Comment by TheOpponent** [comment]
@shrik3 I was able to password-protect only the root to access the stream page and no other sublocations by adding `basic_auth` to a `location = /` block, then making a copy of that block without the `basic_auth` directive in `location /`.

```nginx
    location = / {
        auth_basic           "Restricted";
        auth_basic_user_file /etc/apache2/.htpasswd;
        # Proxy directives.
    }

    location / {
        # Same proxy directives.
    }

```

**Comment by gabek** [comment]
It'd be cool to add a plugin hook so people can build custom auth plugins. I'm not sure what that would look like offhand, but it seems like it could be a nice solution so people can auth against whatever they want.

**Comment by gabek** [comment]
Custom auth plugins will be available in `v0.3.0` and should address this. A simple plugin with a single shared password would be easily possible if that's something a user chose to install.
